src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -0,0 +1,209 @@
+import type { Connection } from '../../connection';
+import type { MongoCredentials } from '../mongo_credentials';
+import type { OIDCMechanismServerStep1, OIDCRequestTokenResult } from '../mongodb_oidc';
+import { AuthMechanism } from '../providers';
+import { TokenEntryCache } from './token_entry_cache';
+import type { Workflow } from './workflow';
+
+/* 5 minutes in milliseconds */
+const TIMEOUT = 300000;
+
+/**
+ * OIDC implementation of a callback based workflow.
+ * @internal
+ */
+export class CallbackWorkflow implements Workflow {
+  cache: TokenEntryCache;
+
+  /**
+   * Instantiate the workflow
+   */
+  constructor() {
+    this.cache = new TokenEntryCache();
+  }
+
+  /**
+   * Get the document to add for speculative authentication. Is empty when
+   * callbacks are in play.
+   */
+  speculativeAuth(): Promise<Document> {
+    return Promise.resolve({});
+  }
+
+  /**
+   * Execute the workflow.
+   *
+   * Steps:
+   * - If an entry is in the cache
+   *   - If it is not expired
+   *     - Skip step one and use the entry to execute step two.
+   *   - If it is expired
+   *     - If the refresh callback exists
+   *       - remove expired entry from cache
+   *       - call the refresh callback.
+   *       - put the new entry in the cache.
+   *       - execute step two.
+   *     - If the refresh callback does not exist.
+   *       - remove expired entry from cache
+   *       - call the request callback.
+   *       - put the new entry in the cache.
+   *       - execute step two.
+   * - If no entry is in the cache.
+   *   - execute step one.
+   *   - call the refresh callback.
+   *   - put the new entry in the cache.
+   *   - execute step two.
+   */
+  async execute(connection: Connection, credentials: MongoCredentials): Promise<Document> {
+    const entry = this.cache.getEntry(connection.address, credentials.username);
+    if (entry) {
+      // Check if the entry is not expired.
+      if (entry.isValid()) {
+        // Skip step one and execute the step two saslContinue.
+        try {
+          const result = await finishAuth(entry.tokenResult, undefined, connection, credentials);
+          return result;
+        } catch (error) {
+          // If authentication errors when using a cached token we remove it from
+          // the cache.
+          this.cache.deleteEntry(connection.address, credentials.username);
+          return Promise.reject(error);
+        }
+      } else {
+        // Remove the expired entry from the cache.
+        this.cache.deleteEntry(connection.address, credentials.username);
+        // Execute a refresh of the token and finish auth.
+        return this.refreshAndFinish(
+          connection,
+          credentials,
+          entry.serverResult,
+          entry.tokenResult
+        );
+      }
+    } else {
+      // No entry means to start with the step one saslStart.
+      const executeCommand = promisify(connection.command.bind(connection));
```

Comment:
Can we consider using `promisify` in the Connection constructor instead of per-invocation?  See https://github.com/mongodb/node-mongodb-native/blob/a45e0e1fc74a03ea43eb8cff058d758c3590b820/src/sdam/topology.ts#L235

Storing the promisified version makes the cost of using promisify once per-connection, instead of each time we execute a command from this module